### DISCUSSION TITLE
`update` method for `useFieldArray`

### DISCUSSION THREAD

**Discussion by bluebill1049** [react-hook-form/react-hook-form#5832] [RFC]
# Context

The current API to update the field array is using `setValue`, which can b performed by the following API.

```tsx
setValue('fieldArray', [
  ...updatedFieldArrayValue
])

setValue('fieldArray.0.test', data)
```

The setValue method then needs to look up the fields reference to identify its field array inputs or normal inputs. 

## Problem

However, the current `setValue` API doesn't actually re-render the `useFieldArray` which causes confusion when developers assume it will behave similarly to the rest of useFieldArray's API, such as `append`, `prepend` and etc.

-------

# Feature request

https://github.com/react-hook-form/react-hook-form/issues/5728

-------

# Proposal

Ideally, we should have an update function to update the field array and trigger a re-render to the `userFieldArray`.

```tsx
const { update } = useFieldArray();

update(0, data); // update an individual field array node
```

```tsx
// it possible to call multiple time as well
update(0, data);
update(1, data);
```

the action itself will trigger a re-render to the entire field array and update all its references.

------

## PR

https://github.com/react-hook-form/react-hook-form/pull/5835

**Comment by jhg023** [comment]
Being able to update an individual field array node using its index will be very useful to me (and presumably many others). Thank you!

**Reply by bluebill1049** [reply]
Thanks for sending the feature request as well.

**Comment by thanh-nguyen-95** [comment]
It's a huge yes for me. I'm looking forward to it.

Btw, there's a missing `)` at the end of `update([...fieldArrayData])`.

**Reply by bluebill1049** [reply]
Thanks 👍 will get that sorted.

**Comment by bluebill1049** [comment]
cc @kotarella1110 @jorisre

**Reply by jorisre** [reply]
💯  agree

**Reply by kotarella1110** [reply]
This is great feature 🎉
agree 👍

**Comment by wdfinch** [comment]
This is great. I use `setValue` all the time to change field array elements and this makes it so much easier.

**Comment by wdfinch** [comment]
In order to use this method, I'm assuming that all elements of the field array need to be mounted correct?

**Reply by bluebill1049** [reply]
I will write a test and let you know. I don't think we need to mount the inputs.

**Comment by behzadmehrabi** [comment]
thanks man, I'm having the exact same issue :))

**Comment by bluebill1049** [comment]
This feature is released. 7.11.0 you can refer to the release note and change log.

**Reply by jhg023** [reply]
Thank you again for prioritizing this feature. I just tested the latest version and it fixed the initial issue that I faced (which was what prompted me to create the feature request in the first place).

**Comment by Mana-COMO** [comment]
I would be much appreciated if you could show me how to use this {update} code sample🙏🙏🙏

**Comment by cbdeveloper** [comment]
What if there are async updates (and the indexes could change during the async cycle) ?
Can't we update by id?

https://github.com/react-hook-form/react-hook-form/discussions/8151

**Reply by Dragomir-Ivanov** [reply]
Then you find that index you need!

The way I do with current RHF:
```
  const { fields, append, remove, update } = useFieldArray({
    ......
    keyName: "_keyId",
  })

....
const handleClose = useCallback(
    (asset: Asset | undefined) => {
      if (asset) {
        if (asset.id) {
          const idx = fields.findIndex((f) => f.id! === asset.id)
          if (idx !== -1) {
            update(idx, asset)
          }
        } else {
          asset.id = generateId()
          append(asset)
        }
      }
    },
    [fields, append, update]
  )
```

#8151

**Comment by sahil-1912** [comment]
How can i use field array id for setting different rendering condition for rendering the element different way?
To be precise i got nested form and depending on the drop down in main array i want to show "add more option" from nested array.

**Reply by leapful** [reply]
@sahil-1912 

Please open a new discussion as your question might be not related to this discussion.
